### ISSUE TITLE
Twirp Ruby doesn't ignore unknown JSON Fields

### ISSUE THREAD

**Issue by alusco-scratch** [twitchtv/twirp-ruby#59]
In [this commit](https://github.com/protocolbuffers/protobuf/commit/aa5c12e8828b3f2f7f52e92a4ec7173d5fb0804f#diff-f0bdd14748207b0a0b01636e573ea5da8686a405d9f8a733f35bd8bd5c49b2c9) the protobuf library added the flag "ignore_unknown_fields" to the decode_json method.  In an ideal world twirp ruby's Encoding.decode method should set this flag when decoding JSON to make it more like protobuf.

I'm happy to make this change but there's a few ways to implement it based on your preferences so wanted to get your thoughts before making a PR.

## My preferred way
This change came out 2 years ago in version 3.6.1.  I don't think it's a huge lift to just bump the minimum version of protobuf required given how old the change is. To do this we'd just:
* Change the code to pass the flag in Encoding: `when JSON  then msg_class.decode_json(bytes, ignore_unknown_fields: true`
* Bump the minimum required protobuf version to >=3.6.1 which is where they pushed it.

## Make it a flag of some kind when instantiating the service
This option would be fully backwards compatible and require no version bump of protobufs but we'd have to change the service some.  A few options could be:
* Change service to take in an options object that has encoding flags.  Pass options to the encoding object to specify behavior.
* Bigger change where we make a custom encoding class able to be passed in or provide a hook to override encoding/decoding (might be nice but possibly too much power)

## Make it so the client can specify it wants the behavior
This is also backwards compatible but requires fudging the Twirp spec at least in some way.  A few options:
* We use the parameter field of content types to add something like 'application/json; ignore_unknown_fields=true'.  This is a valid content type according to the [rfc](https://www.w3.org/Protocols/rfc1341/4_Content-Type.html).  I have no idea if this is common practice but it is pretty neat and probably the least offensive way for us to make this client controllable.
* We allow specifying a header or URL parameter to specify it.  Header might get blocked by proxies but is a pretty easy thing to support in the service.  A url parameter may go against the twirp spec but generally feels sorta gross and against the spirit.
* Could violate the twirp spec and add a new content type for it 'application/x-json-whatever'.  This is a bad option but here for completeness.

Let me know if you have a preference.

Also separately let me know how open you are to some other PRs or what process would be best to chat about them.  We use Twirp extensively at Scratch and have a lot of libraries or some support for things I'd love to add upstream. A few of the major candidates:
1. I'd love to upstream some deadline stuff.  The twirp spec is a bit light on this but we have some middleware that propagates deadlines and in an ideal world the twirp plugin could parse a default deadline as an option from the RPC service and add it to the rpcdef.  This would be a bigger effort but a lot of value if you're open to it.
2. A framework for testing that allows mounting the service using rack in a unit test and making calls (this one is awesome but will require decoupling a bit from our framework).  Also some rspec matchers for be_twirp_error, be_twirp_ok, etc...
3. A framework for defining handlers as individual classes, mounting multiple services, etc..  Probably hard to upstream but could add them to a contrib (if not I may just make a standalone gem for it someday)
4. Some extension methods that do things like client.mymethod!(whatever) that raises if there's a twirp error or returns resp.data directly -- I realize you reverted a PR adding a TwirpException earlier which make sense but I do like the bang version of the methods raising since it removes a lot of unit test boiler plate.

**Comment by 3ventic** [comment]
Your preferred way sounds reasonable and ignoring unknown fields is in line with the Go implementation.

As a side note, it seems to have been released with [3.7.0](https://github.com/protocolbuffers/protobuf/releases/v3.7.0), but the bump there is acceptable nevertheless.

**Comment by alusco-scratch** [comment]
Ah good catch, I pulled the version sorta quick :)  I'll send a quick PR for this then.

Any appetite for the deadline stuff or would you like my to open a new issue and describe the approach?

**Comment by marioizquierdo** [comment]
Thank you for the contribution!
The preferred way sounds good to me.
Open a PR and, I'll look at it as soon as it is ready. Thanks!

**Comment by 3ventic** [comment]
> Any appetite for the deadline stuff or would you like my to open a new issue and describe the approach?

A separate issue would be better. If it involves changes to the twirp specification, it would be best in the main repository.

**Comment by alusco-scratch** [comment]
> > Any appetite for the deadline stuff or would you like my to open a new issue and describe the approach?
> 
> A separate issue would be better. If it involves changes to the twirp specification, it would be best in the main repository.

I'll open a separate issue in a bit, it doesn't involve changing the twirp specification but would extend the parser to parse a deadline option when generating service code.

**Comment by marioizquierdo** [comment]
> would extend the parser to parse a deadline option when generating service code
Twirp is not kind on using options on the code-generation phase if that is what you mean. That is because they add operational complexity (e.g. if you share a proto file and other team is generating the clients from it). An alternative for customization is adding options on the service constructor, which can be done in the twirp gem (would require upgrading the Ruby gem, not the generator plugin). In the Go implementation, there are serverOptions and clientOptions that are used for this reason.
